From 1e305db19b7d480f40ea2871ffd75916cd4c7d69 Mon Sep 17 00:00:00 2001
From: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
Date: Fri, 17 Apr 2026 14:01:41 +0530
Subject: [PATCH] drivers: remoteproc: q6v5_pas: Add Break-at-start support

Add break-at-start debug option using module param "debug_q6". When this
is set, TCSR_SPARE_REG0 is set to 1. PAS provider will read this
register and configure the Q6 debug config to break-at-start so that
JTAG can be attached and any early issues can be debugged.

Change-Id: I4b16d065b8d28f559079b14950a3b0b937e5ab38
Signed-off-by: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
---
 drivers/remoteproc/qcom_q6v5_pas.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/remoteproc/qcom_q6v5_pas.c b/drivers/remoteproc/qcom_q6v5_pas.c
index a9ea0c2348fa..56b5eb5c5f20 100644
--- a/drivers/remoteproc/qcom_q6v5_pas.c
+++ b/drivers/remoteproc/qcom_q6v5_pas.c
@@ -34,6 +34,12 @@
 
 #define ADSP_DECRYPT_SHUTDOWN_DELAY_MS	100
 
+#define TCSR_SPARE_REG0		0x1959000
+#define TCSR_SPARE_REG0_SIZE	4
+
+static int debug_q6;
+module_param(debug_q6, int, 0644);
+
 struct adsp_data {
 	int crash_reason_smem;
 	const char *firmware_name;
@@ -248,6 +254,7 @@ static int adsp_load(struct rproc *rproc, const struct firmware *fw)
 static int adsp_start(struct rproc *rproc)
 {
 	struct qcom_adsp *adsp = rproc->priv;
+	void *debug_q6_reg;
 	int ret;
 
 	ret = qcom_q6v5_prepare(&adsp->q6v5);
@@ -306,6 +313,19 @@ static int adsp_start(struct rproc *rproc)
 
 	qcom_pil_info_store(adsp->info_name, adsp->mem_phys, adsp->mem_size);
 
+	if (debug_q6) {
+		debug_q6_reg = ioremap(TCSR_SPARE_REG0, TCSR_SPARE_REG0_SIZE);
+		if (!debug_q6_reg) {
+			dev_err(adsp->dev, "Failed to ioremap debug_q6 register\n");
+			ret = PTR_ERR(debug_q6_reg);
+			goto release_pas_metadata;
+		}
+
+		writel(0x1, debug_q6_reg);
+		dev_info(adsp->dev, "Writing 1 to TCSR_SPARE_REG0\n");
+		iounmap(debug_q6_reg);
+	}
+
 	ret = qcom_pas_auth_and_reset(adsp->pas_id);
 	if (ret) {
 		dev_err(adsp->dev,
@@ -313,8 +333,12 @@ static int adsp_start(struct rproc *rproc)
 		goto release_pas_metadata;
 	}
 
+wait_for_start:
 	ret = qcom_q6v5_wait_for_start(&adsp->q6v5, msecs_to_jiffies(5000));
 	if (ret == -ETIMEDOUT) {
+		if (debug_q6)
+			goto wait_for_start;
+
 		dev_err(adsp->dev, "start timed out\n");
 		qcom_pas_shutdown(adsp->pas_id);
 		goto release_pas_metadata;
-- 
2.34.1

